src/stan/mcmc/chainset.hpp

```diff
@@ -0,0 +1,567 @@
+   * diagnostic for a set of per-chain draws, for bulk and tail Rhat.
+   * Based on paper https://arxiv.org/abs/1903.08008
+   *
+   * @param name parameter name
+   * @return pair (bulk_rhat, tail_rhat)
+   */
+  std::pair<double, double> split_rank_normalized_rhat(
+      const std::string& name) const {
+    return split_rank_normalized_rhat(index(name));
+  }
+
+  /**
+   * Computes the effective sample size (ESS) for the specified
+   * parameter across all chains, according to the algorithm presented in
+   * https://arxiv.org/abs/1903.08008, section 3.2 for folded (split)
+   * rank-normalized ESS for both bulk and tail ESS.
+   *
+   * @param index parameter index
+   * @return pair (bulk_ess, tail_ess)
+   */
+  std::pair<double, double> split_rank_normalized_ess(const int index) const {
+    return analyze::split_rank_normalized_ess(samples(index));
+  }
+
+  /**
+   * Computes the effective sample size (ESS) for the specified
+   * parameter across all chains, according to the algorithm presented in
+   * https://arxiv.org/abs/1903.08008, section 3.2 for folded (split)
+   * rank-normalized ESS for both bulk and tail ESS.
+   *
+   * @param name parameter name
+   * @return pair (bulk_ess, tail_ess)
+   */
+  std::pair<double, double> split_rank_normalized_ess(
+      const std::string& name) const {
+    return split_rank_normalized_ess(index(name));
+  }
+
+  /**
+   * Computes the mean Monte Carlo error estimate for the central 90% interval.
+   * See https://arxiv.org/abs/1903.08008, section 4.4.
+   * Follows implementation in the R posterior package
+   *
+   * @param index parameter index
+   * @return pair (bulk_ess, tail_ess)
+   */
+  double mcse_mean(const int index) const {
+    double ess_bulk = analyze::split_rank_normalized_ess(samples(index)).first;
+    return sd(index) / std::sqrt(ess_bulk);
+  }
+
+  /**
+   * Computes the mean Monte Carlo error estimate for the central 90% interval.
+   * See https://arxiv.org/abs/1903.08008, section 4.4.
+   * Follows implementation in the R posterior package.
+   *
+   * @param name parameter name
+   * @return pair (bulk_ess, tail_ess)
+   */
+  double mcse_mean(const std::string& name) const {
+    return mcse_mean(index(name));
+  }
+
+  /**
+   * Computes the standard deviation of the Monte Carlo error estimate
+   * https://arxiv.org/abs/1903.08008, section 4.4.
+   * Follows implementation in the R posterior package.
+   *
+   * @param index parameter index
+   * @return pair (bulk_ess, tail_ess)
+   */
+  double mcse_sd(const int index) const {
+    Eigen::MatrixXd s = samples(index);
+    Eigen::MatrixXd s2 = s.array().square();
+    double ess_s = analyze::split_rank_normalized_ess(s).first;
+    double ess_s2 = analyze::split_rank_normalized_ess(s2).first;
+    double ess_sd = std::min(ess_s, ess_s2);
+    return sd(index)
+           * std::sqrt(std::exp(1) * std::pow(1 - 1 / ess_sd, ess_sd - 1) - 1);
```

Comment:
We can use `e()` from Stan Math instead of exp(1)
https://github.com/stan-dev/math/blob/9b1f08be7e76c6468fc2fcf2f75d641a619ae349/stan/math/prim/fun/constants.hpp#L20C24-L20C29

Reply:
the new summary will include MAD, and ESS_bulk and ESS_tail, and won't include ESS/second; we will need to document all of this, as well as the fact that quantiles are calculated using different libraries than before.